File: iOS/swiftui-mfa/MfaSample/MfaSample/Core/Services/DiagnosticLogger.swift
//
//  DiagnosticLogger.swift
//  MfaSample
//
//  Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
//
//  This software may be modified and distributed under the terms
//  of the MIT license. See the LICENSE file for details.
//

import Foundation
import Combine
import os.log

/// Log entry representing a single diagnostic log message.
struct LogEntry: Identifiable {
    let id = UUID()
    let timestamp: Date
    let level: LogLevel
    let category: String
    let message: String

    enum LogLevel: String {
        case debug = "DEBUG"
        case info = "INFO"
        case warning = "WARNING"
        case error = "ERROR"

        var icon: String {
            switch self {
            case .debug: return "ladybug"
            case .info: return "info.circle"
            case .warning: return "exclamationmark.triangle"
            case .error: return "xmark.circle"
            }
        }
    }
}

/// Service for capturing and managing diagnostic logs.
@MainActor
class DiagnosticLogger: ObservableObject {
    static let shared = DiagnosticLogger()

    // MARK: - Published State
    @Published private(set) var logs: [LogEntry] = []

    // MARK: - Properties
    private let maxLogEntries = 1000
    private let logger = Logger(subsystem: Bundle.main.bundleIdentifier ?? "com.pingidentity.mfaexample", category: "Diagnostic")

    // MARK: - Initialization
    private init() {}

    // MARK: - Logging Methods
    /// Logs a debug message
    func debug(_ message: String, category: String = "General") {
        log(message, level: .debug, category: category)
        logger.debug("\(category): \(message)")
    }

    /// Logs an info message
    func info(_ message: String, category: String = "General") {
        log(message, level: .info, category: category)
        logger.info("\(category): \(message)")
    }

    /// Logs a warning message
    func warning(_ message: String, category: String = "General") {
        log(message, level: .warning, category: category)
        logger.warning("\(category): \(message)")
    }

    /// Logs an error message
    func error(_ message: String, category: String = "General") {
        log(message, level: .error, category: category)
        logger.error("\(category): \(message)")
    }

    /// Internal logging method
    private func log(_ message: String, level: LogEntry.LogLevel, category: String) {
        guard UserPreferences.shared.diagnosticLogging else {
            return
        }

        let entry = LogEntry(
            timestamp: Date(),
            level: level,
            category: category,
            message: message
        )

        logs.append(entry)

        // Trim logs if exceeding max entries
        if logs.count > maxLogEntries {
            logs.removeFirst(logs.count - maxLogEntries)
        }
    }

    // MARK: - Export Methods
    /// Exports all logs as a formatted string
    func exportLogs() -> String {
        let dateFormatter = DateFormatter()
        dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss.SSS"

        var output = "MFA Example - Diagnostic Logs\n"
        output += "Generated: \(dateFormatter.string(from: Date()))\n"
        output += "Total Entries: \(logs.count)\n"
        output += String(repeating: "=", count: 80) + "\n\n"

        for entry in logs {
            let timestamp = dateFormatter.string(from: entry.timestamp)
            output += "[\(timestamp)] [\(entry.level.rawValue)] [\(entry.category)]\n"
            output += "\(entry.message)\n\n"
        }

        return output
    }

    /// Clears all logs
    func clearLogs() {
        logs.removeAll()
        info("Logs cleared", category: "DiagnosticLogger")
    }

    /// Gets logs filtered by level
    func logs(withLevel level: LogEntry.LogLevel) -> [LogEntry] {
        logs.filter { $0.level == level }
    }

    /// Gets logs filtered by category
    func logs(withCategory category: String) -> [LogEntry] {
        logs.filter { $0.category == category }
    }
}

File: android/kotlin-authenticatorapp/app/src/main/kotlin/com/pingidentity/authenticatorapp/data/DiagnosticLogger.kt
/*
 * Copyright (c) 2025 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

package com.pingidentity.authenticatorapp.data

import android.annotation.SuppressLint
import com.pingidentity.logger.Logger
import com.pingidentity.logger.Standard
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.UUID.randomUUID
import java.util.concurrent.ConcurrentLinkedQueue

/**
 * Data class representing a log entry.
 */
data class LogEntry(
    val id: String = randomUUID().toString(),
    val timestamp: String,
    val level: String,
    val message: String,
    val throwable: String? = null
)

/**
 * Diagnostic logger that captures logs in memory for debugging purposes.
 * This logger wraps the standard logger and also stores logs for later viewing.
 */
object DiagnosticLogger : Logger {
    private val standardLogger = Standard()
    private val logEntries = ConcurrentLinkedQueue<LogEntry>()

    @SuppressLint("ConstantLocale")
    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS", Locale.getDefault())
    
    private const val MAX_LOG_ENTRIES = 1000

    private val _logs = MutableStateFlow<List<LogEntry>>(emptyList())
    val logs: StateFlow<List<LogEntry>> = _logs.asStateFlow()
    
    private fun addLogEntry(level: String, message: String, throwable: Throwable? = null) {
        val timestamp = dateFormat.format(Date())
        val throwableString = throwable?.let { 
            "${it.javaClass.simpleName}: ${it.message}\n${it.stackTraceToString()}"
        }
        
        val logEntry = LogEntry(
            timestamp = timestamp,
            level = level,
            message = message,
            throwable = throwableString
        )
        
        logEntries.add(logEntry)
        
        // Keep only the last MAX_LOG_ENTRIES entries
        while (logEntries.size > MAX_LOG_ENTRIES) {
            logEntries.poll()
        }
        
        // Update the StateFlow
        _logs.value = logEntries.toList()
    }
    
    override fun d(message: String) {
        standardLogger.d(message)
        addLogEntry("DEBUG", message)
    }
    
    override fun i(message: String) {
        standardLogger.i(message)
        addLogEntry("INFO", message)
    }
    
    override fun w(message: String, throwable: Throwable?) {
        standardLogger.w(message, throwable)
        addLogEntry("WARN", message, throwable)
    }
    
    override fun e(message: String, throwable: Throwable?) {
        standardLogger.e(message, throwable)
        addLogEntry("ERROR", message, throwable)
    }
    
    /**
     * Clear all captured log entries.
     */
    fun clearLogs() {
        logEntries.clear()
        _logs.value = emptyList()
    }
    
    /**
     * Export all logs as a formatted string.
     */
    fun exportLogs(): String {
        val sb = StringBuilder()
        sb.appendLine("=== Diagnostic Logs Export ===")
        sb.appendLine("Exported at: ${dateFormat.format(Date())}")
        sb.appendLine("Total entries: ${logEntries.size}")
        sb.appendLine()
        
        logEntries.forEach { entry ->
            sb.appendLine("[${entry.timestamp}] ${entry.level}: ${entry.message}")
            entry.throwable?.let { throwable ->
                sb.appendLine("Exception: $throwable")
            }
            sb.appendLine()
        }
        
        return sb.toString()
    }
}